subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskOutputCacheCommandFactory.java

```diff
@@ -103,16 +125,45 @@ public BuildCacheKey getKey() {
             return new BuildCacheLoadCommand.Result<TaskOutputOriginMetadata>() {
                 @Override
                 public long getArtifactEntryCount() {
-                    return unpackResult.entries;
+                    return unpackResult.getEntries();
                 }
 
                 @Override
                 public TaskOutputOriginMetadata getMetadata() {
-                    return unpackResult.originMetadata;
+                    return unpackResult.getOriginMetadata();
                 }
             };
         }
 
+        private void updateSnapshots(ImmutableListMultimap<String, FileSnapshot> propertiesFileSnapshots) {
+            ImmutableMap<String, ResolvedTaskOutputFilePropertySpec> propertySpecsMap = Maps.uniqueIndex(outputProperties, new Function<TaskOutputFilePropertySpec, String>() {
+                @Override
+                public String apply(TaskOutputFilePropertySpec propertySpec) {
+                    return propertySpec.getPropertyName();
+                }
+            });
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> propertySnapshotsBuilder = ImmutableSortedMap.naturalOrder();
+            for (String propertyName : propertiesFileSnapshots.keySet()) {
+                ResolvedTaskOutputFilePropertySpec property = propertySpecsMap.get(propertyName);
+                List<FileSnapshot> fileSnapshots = propertiesFileSnapshots.get(propertyName);
+                FileCollectionSnapshotBuilder builder = new FileCollectionSnapshotBuilder(UNORDERED, OutputPathNormalizationStrategy.getInstance(), stringInterner);
+                builder.visitFileTreeSnapshot(fileSnapshots);
+
+                propertySnapshotsBuilder.put(propertyName, builder.build());
+
+                if (fileSnapshots.size() == 1) {
+                    FileSnapshot singleSnapshot = fileSnapshots.get(0);
+                    if (singleSnapshot.getType() != FileType.RegularFile) {
+                        throw new IllegalStateException(String.format("Only a regular file should be produced by unpacking property '%s', but saw a %s", propertyName, singleSnapshot.getType()));
+                    }
+                    fileSystemMirror.putFile(singleSnapshot);
+                } else {
+                    fileSystemMirror.putDirectory(new DirectoryTreeDetails(property.getOutputFile().getAbsolutePath(), fileSnapshots));
```

Comment:
Yeah, there should be no chance to collide with someone else snapshotting our just-unpacked outputs. It would be nice to express this mutual exclusion in a more explicit way, though.